### ISSUE TITLE
[GHA][EC2] Instance provisioned for load testing

### ISSUE THREAD

**Issue by oseasmoran73** [department-of-veterans-affairs/va.gov-team#40552]
## Description

Health check-in team would like to create an instance for a few months to run some load testing with locust for an app. The goal would be to create a minimal AMI with `docker` and create a Github Action workflow to allow other VFS teams to use as well

## Background/context 
[Slack](https://dsva.slack.com/archives/CBU0KDSB1/p1650907552995939)

- Lives in vets-api already

> if you can get a Runner to 1) clone a repo, 2) docker run locust with the repo mounted, and 3) store the report externally (since the Runner is ephemeral), then we're pretty much good to go

## Technical notes

- Set up a GHA workflow that allows a few parameters, then runs all of it in a Runner
  - Runner = EC2 w/ Docker installed
- Locust would need to run headless and allow few inputs (such as `-u` or `-r`) & path-to-locust-file
- Store reports (TBD)

---
## Tasks
- [x] Investigate [locust](https://docs.locust.io/en/stable/running-without-web-ui.html) headless & documentation of parameters (must run in a [container](https://docs.locust.io/en/1.5.2/running-locust-docker.html))
- [x] Create basic AMI
- [x] Create Github Action workflow to accept parameters
- [x] Request testing from users
- [x] Evaluate if this needs to be it's own ASG
- [x] Ensure Runners are labeled correctly and do not interfere with existing workflows

## Acceptance Criteria 
- [x] Dynamically create EC2 instance for load testing in AWS

**Comment by oseasmoran73** [comment]
Couple of things that we may need to consider: 

- Running the headless state will allow the flow to continue. That being said, it will immediately proceed to the next step and shut down the EC2 (I presume, have not actually tested). We may need to add dynamically a variable for `run_time`
  - Wonder if we should have a default for a specific time?

- Should the directory of the locust_file_path also contain the requirements for installation? 
  - e.x: if locust_file_path = `/pl/testing/locust.py`, `/pl/testing/requirement.txt` lives in this path

**Comment by oseasmoran73** [comment]
@ph-One Would like your input on the follow questions above

**Comment by oseasmoran73** [comment]
As of now I have the the following:

![image](https://user-images.githubusercontent.com/10648480/167452728-1268ccb8-a7c9-44f9-b59f-6808c4bfb772.png)

Note: When adding path to locust file, does not need to start with leading `/` (e.x: `loadtest/vets-website/locust.py` instead of `/loadtest/vets-website/locust.py`)

Note: The assumption is that there is a `requirement.txt` at the root of the of repo checked out to run `pip3`

**Comment by oseasmoran73** [comment]
Awaiting feedback from health check-in team

**Comment by mleclerc00** [comment]
> Couple of things that we may need to consider:
> 
> * Running the headless state will allow the flow to continue. That being said, it will immediately proceed to the next step and shut down the EC2 (I presume, have not actually tested). We may need to add dynamically a variable for `run_time`
>   
>   * Wonder if we should have a default for a specific time?
> * Should the directory of the locust_file_path also contain the requirements for installation?
>   
>   * e.x: if locust_file_path = `/pl/testing/locust.py`, `/pl/testing/requirement.txt` lives in this path

Regarding the runner shutting down prior to the tests being completed, could we just throw a `needs` condition on it actually shutting down? 

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idneeds

Essentially preventing it from moving onto that job until the tests either pass, fail, or timeout?

**Comment by oseasmoran73** [comment]
>Essentially preventing it from moving onto that job until the tests either pass, fail, or timeout?

Added a couple of flags to allow the runner to persist. It will persist for a max of one hour. We can probably add a dynamic variable to longer TTL, but waiting on feedback before implementing it

**Comment by randomsync** [comment]
Thanks @oseasmoran73 I'm working on some locust scripts that'll help me test drive this. Will let you know.

**Comment by kanchanasuriya** [comment]
This github action seems to require `requirements.txt` but our load test script doesn't need any additional modules to install through pip3.

```
Run pip3 install -r ./requirements.txt
ERROR: Could not open requirements file: [Errno 2] No such file or directory: './requirements.txt'
Error: Process completed with exit code 1.
```
Reference Link: https://github.com/department-of-veterans-affairs/devops/runs/6497735224?check_suite_focus=true

**Comment by kanchanasuriya** [comment]
@oseasmoran73 Used devops as the repository for running the script & passed in `40651_check-in_load_test_scripts` branch  in the workflow

**Comment by oseasmoran73** [comment]
I took a look at the branch and it was difficult to see what was being tested as it is out of sync with master.

What you can do for now is create a `requirement.txt` folder at the root level and test for now.

See [demo](https://github.com/department-of-veterans-affairs/devops/compare/master...demo/gha/performance-load-testing) branch I created

**Comment by kanchanasuriya** [comment]
We tried that as well @oseasmoran73.  Created requirement.txt in [check_in/](https://github.com/department-of-veterans-affairs/devops/tree/40651_check-in_load_test_scripts/loadtest/check_in) with `locust.py` in [check_in/check_in](https://github.com/department-of-veterans-affairs/devops/tree/40651_check-in_load_test_scripts/loadtest/check_in/check-in).  Workflow continues to fail looking `requirment.txt` in current path

**Comment by oseasmoran73** [comment]
You created in `devops/check_in/requirements.txt` rather than `/devops/requirements.txt`. Add it to the root folder.

Most likely, we will need to have a way to identify where `requirements.txt` lives. Give it a whirl and see if that works.

**Comment by kanchanasuriya** [comment]
No luck @oseasmoran73 .  Pushed `requirements.txt` in [root/](https://github.com/department-of-veterans-affairs/devops/tree/40651_check-in_load_test_scripts). Still workflow fails in `Install dependencies` step with `ERROR: Could not open requirements file:`

**Comment by oseasmoran73** [comment]
@kanchanasuriya Can you sync your branch with the latest master so I can look at the files committed only? I will give it a try on my end once that is done

**Comment by kanchanasuriya** [comment]
Rebased with master and pushed the changes to `40651_check-in_load_test_scripts` branch.  GitHub action still fails with `ERROR: Could not open requirements file:`

**Comment by oseasmoran73** [comment]
I got your branch working, so I presume there is a misalignment on what was being entered vs what we have documented. I will be sure to update our document

Here is a [workflow](https://github.com/department-of-veterans-affairs/devops/actions/runs/2373296901) that installs dependency. It doesn't quite succeed as it looks like its trying to get something from python2

![image](https://user-images.githubusercontent.com/10648480/169887427-fdb3b7a0-2b10-4c02-9d51-1aec9e8d1c02.png)

**Comment by kanchanasuriya** [comment]
Glad to see it moved forward @oseasmoran73.  Now it fails with one of the dependency package installation,

```
Collecting subprocess32==3.2.7
  Downloading subprocess32-3.2.7.tar.gz (54 kB)
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ta8lt134/subprocess32/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ta8lt134/subprocess32/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-ta8lt134/subprocess32/pip-egg-info
         cwd: /tmp/pip-install-ta8lt134/subprocess32/
    Complete output (1 lines):
    This backport is for Python 2.x only.
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
Error: Process completed with exit code 1.
```

**Comment by oseasmoran73** [comment]
Was chatting with @kanchanasuriya on the side trying to debug. Asked her to add `setup.py` prior to executing `subprocess32` and had no success. Will investigate from our side and most likely need to upgrade pip3 prior to install dependencies

**Comment by oseasmoran73** [comment]
Hi @kanchanasuriya 

Is subprocess32 required? I am wondering if we were to omit (since we are on python3) if we can get around it for the time being. Let me know what your thoughts on it while I still investigate further

**Comment by oseasmoran73** [comment]
hi @kanchanasuriya 

I am still in the process of debugging! Sorry that this is taking longer than usual.

I have a question, when installed locally, are there any conflicts with the `requirements.txt`? I ask because I am currently purging some of the depdencies and there seems to be some conflict

See [workflow](https://github.com/department-of-veterans-affairs/devops/runs/6693994915?check_suite_focus=true) below:

```
The conflict is caused by:
    The user requested pyzmq==17.0.0
    locust 2.5.1 depends on pyzmq>=22.2.1
```

**Comment by oseasmoran73** [comment]
Tested the final [workflow](https://github.com/department-of-veterans-affairs/devops/runs/6695448121?check_suite_focus=true) here and successfully was able to get it up and running

I omitted `subprocess` and `setuptools` from the `requirements.txt` and left everything else the same.

As mentioned, I think there is version conflicts with the pinned versions. I will leave that up for your team to resolve and get this PR merged soon to unblock y'all. Thanks for being patient!

**Comment by kanchanasuriya** [comment]
Pls, note, that slightly switching gears on using the [K6](https://k6.io/docs/getting-started/installation/) for running load test.  Would be great to have K6 installed in the provisioned host.

**Comment by oseasmoran73** [comment]
@kanchanasuriya This is currently set up for locust. If it is being changed to k6, we will need to go back to the "drawing board" and think of it how to make it more flexible to different load tests

**Comment by oseasmoran73** [comment]
We were able to get the dependency issue resolved. It looks like when @kanchanasuriya runs the workflow it fails to pull the locust ECR, but for me. Suspect it has to do with permissions, will need to further investigate

**Comment by oseasmoran73** [comment]
Closing issue as health check in team is using a different EC2 instance.

Creating a separate ticket specifically for ECR issue.

Please feel free to reopen should you feel this is incorrect
